### DISCUSSION TITLE
Cannot import component in MDX file.

### DISCUSSION THREAD

**Discussion by supananplus** [vercel/next.js#63178] [Help]
### Summary

This is my MDX file:
```mdx
---
title: 'First Blog Post'
author: 'Hamed Bahram'
publishDate: 'March 4th, 2023'
---
import Button from "../components/button.js";

# This is our first post

Here is a regular paragraph

here is a list:

- Item one
- Item two
- Item three

## This is an h2 element
# This is an h1 element


# MDX + Next.js

_Look, a button!_ 👇

<Button>👋 Hello</Button>
```
This is my button component: 
```js
export default function Button({ children }) {
  return (
    <button
      style={{
        borderRadius: "3px",
        border: "1px solid black",
        color: "black",
        padding: "0.5em 1em",
        cursor: "pointer",
        fontSize: "1.1em",
      }}
    >
      {children}
    </button>
  );
}
```

This is how I call it:
```js
import { getPostBySlug } from "@/lib/mdx";

const getPageContent = async (slug) => {
  const { meta, content } = await getPostBySlug(slug);
  return { meta, content };
};

export async function generateMetadata({ params }) {
  const { meta } = await getPageContent(params.slug);
  return { title: meta.title };
}

// params => dynamic part or segment of the route
const Page = async ({ params, source }) => {
  const { content } = await getPageContent(params.slug);

  return (
    <section className="py-24">
      <div className="container py-4 prose">{content}</div>
    </section>
  );
};

export default Page;
```

and this is my error message:

> Error: Expected component `Button` to be defined: you likely forgot to import, pass, or provide it.
It’s referenced in your code at `21:1-21:26`

Is anyone here face the same problem as me?

ps. I found similar question but the solution is not work for me at all.

**Comment by icyJoseph** [comment]
Hej

I don't have a project setup to test, but a couple of things to troubleshoot:

- Remove the `.js` from the import statement
- Make sure it is `components/button` and not `components/Button`.

**Reply by supananplus** [reply]
Thank you for your reply I've already removed `.js` and I rechecked file name it's `components/button` and the same error still occur

ps. I use this version `"next": "14.1.3"`

**Reply by icyJoseph** [reply]
Mmm odd, I just remembered I had a project on this machine, with MDX Next.js setup, and I have these at the top:

```mdx
import { Code } from "bright";

import { CounterMdx } from "../examples/counter/bright";
import { CounterActionMdx } from "../examples/counter-action/bright";
// and more...

# Hello World
```

And it just works, though I have that as `page.mdx`

**Reply by supananplus** [reply]
or is it related to this function? I'm not sure
```js
export const getPostBySlug = async (slug) => {
  const realSlug = slug.replace(/\.mdx$/, "");
  const filePath = path.join(rootDirectory, `${realSlug}.mdx`);
  // read content in mdx file
  const fileContent = fs.readFileSync(filePath, { encoding: "utf8" });

  // passes the content as a source to complieMDX function
  const { frontmatter, content } = await compileMDX({
    source: fileContent,
    //parses the front matter as an object and give the front pattern
    options: { parseFrontmatter: true },
  });

  return { meta: { ...frontmatter, slug: realSlug }, content };
};
```

**Reply by supananplus** [reply]
Hi @icyJoseph from the previous issue I think it solved with using MDXProvider but now I got new error

> TypeError: createContext only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/context-in-server-component

Then I added `"use client"` in this file
```js
// app\(site)\posts\[slug]\page.js
"use client";
import { getPostBySlug } from "@/lib/mdx";
import { MDXProvider } from "@mdx-js/react";
import Button from "@/app/components/Button";

const getPageContent = async (slug) => {
  const { meta, content } = await getPostBySlug(slug);
  return { meta, content };
};

const generateMetadata = async ({ params }) => {
  const { meta } = await getPageContent(params.slug);
  return { title: meta.title };
};

const components = {
  Button,
};

// params => dynamic part or segment of the route
const Page = async ({ params, source }) => {
  const { content } = await getPageContent(params.slug);

  return (
    <section className="py-24">
      <div className="container py-4 prose">
        <MDXProvider components={components}>{content}</MDXProvider>
      </div>
    </section>
  );
};

export default Page;
```

but I got this error instead

> ./lib/mdx/index.js:1:0
Module not found: Can't resolve 'fs'
> 1 | import fs from "fs";
Import trace for requested module:
./app/(site)/posts/[slug]/page.js

I tried some solutions which is add `module.export` in `next.config.js` but it doesn't work. Do you have any suggestion to fix it?

**Reply by icyJoseph** [reply]
Right, you have to use MDXProvider within a `use client` module. Take a look at the first code snippet here:

- https://github.com/vercel/next.js/discussions/50897#discussioncomment-6115603
